WIP: Update README to reflect image-based gadget focus (#3221)
inspektor-gadget/inspektor-gadget: README.md

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/LICENSE)
[![License: GPL v2](https://img.shields.io/badge/License-GPL%20v2-blue.svg)](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/LICENSE-bpf.txt)

Inspektor Gadget is an [eBPF](https://ebpf.io/) framework and set of tools to debug and inspect
containers, Kubernetes clusters and Linux hosts. It manages the packaging, deployment and

> execution of eBPF programs encapsulated in [OCI images](https://opencontainers.org/) called Gadgets.

Author: It's not saying it executes eBPF programs, just that it manages that. So I think that's correct. Configuration is too loaded and should be used more targeted.

TODO: Insert terminal/promo video

## Features

* Build and package eBPF programs into OCI images called Gadgets
* Runs on Kubernetes and Linux hosts
* Export data to observability tools with a simple configuration
* Security mechanism to restrict and lock-down which Gadgets can be run.
* Enrichment: map kernel data to high-level resources like Kubernetes and container runtimes
* Create declarative data collection pipelines
* Post-process eBPF data via [WebAssembly](https://webassembly.org/) modules; using any WASM-supported language 
* Supports many modes of operation; cli, client-server, API, embeddable via Golang library

> ## Quick start

Reviewer: What do you think of #3229 for the core concepts?

> ## Quick start

Author: The convention I'm using is that we have a short description to provide just enough interesting to give a basic understanding then link to the more detailed pages in the docs. Thus, I don't think the README should just be links to the docs. It needs to be informative in it's own right, IMO.

This assumes you have the required Inspektor Gadget tools installed as outlined in the [Installation instructions](#installation).

The following examples use the [trace_open Gadget](/gadgets/trace_open/README.md) which triggers when a file is open on the system.

### Kubernetes 

```bash
kubectl gadget deploy
kubectl gadget run trace_open
```

### Linux

This runs inspektor Gadget locally on 

```bash
ig run trace_open
```

### MacOS or Windows (but not exclusively)

> Run the following on a Linux machine called "my-linux-machine" to make `ig` available to the MacOS or Windows client.

Author: But that doesn't show how to run it on MacOS or Windows which is the aim. How do you propose showing that or, rather, what is our suggested method?

```bash
ig --daemon
```

On the MacOS or Windows machine use the client to connect and run commands on the server.

```bash

> gadgetctl --server=my-linux machine

Author: I know this doesn't work. It's just a placeholder to indicate what I want to show. I'll need a suggestion from one of you folks who have run it.

> gadgetctl --server=my-linux machine

Author: I've updated this. Please verify that it works.

As OCI images that use the same tooling as containers and share the same attributes; shareable, modular, etc.
Gadgets are built using the `ig image build` command.
You can find a growing collection of Gadgets 

See the [Gadget documentation](/gadgets/README.md) for more information.

#### :warning: For versions prior to v0.25

Prior to v0.31.0, Inspektor Gadget shipped gadgets in the binary. As of v0.31.0 these ***built-in*** Gadgets are still available and work as before but their use is discourage as they will be deprecate at some point. We encourage users to use ***image-based*** Gadgets going forward as they provide more features and decouple the eBPF programs from the Inspektor Gadget release process.

### What is enrichment

The data that eBPF collects from the kernel includes no knowledge about Kubernetes, container
runtimes or any other high-level user-space concepts. In order to relate this data to these high-level
concepts and make the eBPF data immediately more understandable, Inspektor Gadget automatically

> uses kernel primitives such as mount namespaces, pids, etc. to infer and which high-level

Reviewer: ```suggestion
uses kernel primitives such as mount namespaces, pids, etc. to infer which high-level
```

concepts they relate to; Kubernetes pods, container names, DNS names, etc. The process of augmenting
the eBPF data with these high-level concepts is called enrichment.

Enrichment flows the other way, too. Inspektor Gadget enables users to do high-performance
in-kernel filtering by only referencing high-level concepts such as Kubernetes pods, container
names, etc.; automatically translating these to the corresponding low-level kernel resources.

> See the enrichment documentation

Author: Yep, looks like we'll have some documentation for that soon which we can link to.

### What is an operator

In Inspektor Gadget, an operator is any part of the framework where an action is taken. Some operators are under-the-hood (loading Gadgets, ) while others are user-exposed (enrichment, filtering, export, etc.) and can be reording and overriding. This allows 

> See the [operator documentation](docs/operators.md) for more information.

Author: Right, an asperational link. But looks like we'll have it soon.

### Further learning

Use the [project documentation](/docs/_index.md) to learn more about:

* [Overview]()
* [Concepts]()
* [Components]()
* [Gadgets]()
* [Reference]()
* [Contributing]()

> We've created a blog series that goes into several topics and concepts in details.

Author: That's the plan. But wanted to show how we should use the blog series and highlight it in the README.

> We've created a blog series that goes into several topics and concepts in details.

Author: I've removed these for now.

* [A new approach to eBPF observability]()
* [Concepts: A basic primer on Gadgets]()
* [Concepts: Advance Gadget usage]()
* [Concepts: Operators in Inspektor Gadget]()
* [Collecting metrics with Inspektor Gadget]()
* [Interactive debugging with Inspektor Gadget]()
* [Using Inspektor Gadget to leverage eBPF in your project]
* [Exploring Inspektor Gadget's security mechanisms]()

## Installation

> Inspektor Gadget can be installed in several different ways depending on your environment and needs.

Reviewer: I'd like to avoid repeating this as much as possible. Keeping these document synchronized will be difficult.

> Inspektor Gadget can be installed in several different ways depending on your environment and needs.

Author: I initially liked this idea and still do, just not for the README. I like the fact that this section shows a simple list of the options. I do think the diagram would be great for the detailed install page.

[![License: GPL v2](https://img.shields.io/badge/License-GPL%20v2-blue.svg)](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/LICENSE-bpf.txt)

Inspektor Gadget is an [eBPF](https://ebpf.io/) framework and set of tools to debug and inspect
containers, Kubernetes clusters and Linux hosts. It manages the packaging, deployment and
execution of eBPF programs encapsulated in [OCI images](https://opencontainers.org/) called Gadgets.

TODO: Insert terminal/promo video

## Features

* Build and package eBPF programs into OCI images called Gadgets
* Targets Kubernetes clusters and Linux hosts
* Collect and export data to observability tools with a simple command (and soon via declarative configuration)
* Security mechanisms to restrict and lock-down which Gadgets can be run
* Automatic [enrichment](#what-is-enrichment): map kernel data to high-level resources like Kubernetes and container runtimes
* Supports [WebAssembly](https://webassembly.org/) modules to post-process data and customize IG [operators](#what-is-an-operator); using any WASM-supported language 
* Supports many modes of operation; cli, client-server, API, embeddable via Golang library

## Quick start

> This assumes you have the required Inspektor Gadget tools installed as outlined in the [Installation instructions](#installation).

Reviewer: Shouldn't we move the installation instructions here? If it's the time somebody is reading it, they'll need for sure install the tools to try the quick start.

Author: I didn't want a half to whole page of installation instructions getting in the way of showing how it works. The installation instruction are in the README and pointed to in the relevant places. I'd like to with this and we can see what feedback we get and adjust.

The following examples use the [trace_open Gadget](/gadgets/trace_open/README.md) which triggers when a file is open on the system.

### Kubernetes 

Deploy and run on a Kubernetes cluster

```bash
kubectl gadget deploy
kubectl gadget run trace_open
```

### Linux

Runs Inspektor Gadget locally on Linux.

```bash
ig run trace_open
```

> ### MacOS or Windows (but not exclusively)

Reviewer: I'm not sure this title is right. Should it be somethin glike "using gadgetctl"? It's also possible to use kubectl-gadget from MacOS / Windows and this part doesn't make it clear.

Author: I know, but that doesn't talk to folks who are wondering how to use IG on Mac or Windows. I'm assuming they don't know `gadgetctl`'s purpose and aren't looking for how to use it but how to use IG itself on those platforms.

Run the following on a Linux machine to make `ig` available to clients.

```bash
ig daemon --host=tcp://0.0.0.0:1234
```

On the MacOS or Windows machine use the `gadgetctl` client to connect and run commands on the server.

```bash

> gadgetctl --remote-address=tcp://$IP:1234
> gadgetctl run trace_open

Reviewer: ```suggestion
gadgetctl run trace_open --remote-address=tcp://$IP:1234
```

```

> ***The above demonstrates the simplest command. To learn how to filter, export, etc. please go consult the documentation for the [run](/docs/guides/run.md) command***.

Reviewer: "go consult" seems fine to me.

 
## Core concepts

TODO: Insert system diagram

### What is a Gadget?

> Gadgets are the central component in the Inspektor Gadget framework. A Gadget is an [OCI image](https://opencontainers.org/) that includes one or more eBPF programs, metadata YAML file and, optionally, WASM modules for post processing and a logo file.

Author: Ah, should remove then or add it. :)

> Gadgets are the central component in the Inspektor Gadget framework. A Gadget is an [OCI image](https://opencontainers.org/) that includes one or more eBPF programs, metadata YAML file and, optionally, WASM modules for post processing and a logo file.

Reviewer: Remove the logo part (it's easier than adding support for it :D)

As OCI images, they use the same tooling as containers and share the same attributes; shareable, modular, deployable, etc.

> Gadgets are built using the [`ig image build`](/docs/gadg) command.

Author: We can fix in the clean up followup commit.

### What is enrichment?

The data that eBPF collects from the kernel includes no knowledge about Kubernetes, container
runtimes or any other high-level user-space concepts. In order to relate this data to these high-level
concepts and make the eBPF data immediately more understandable, Inspektor Gadget automatically
uses kernel primitives such as mount namespaces, pids, etc. to infer which high-level
concepts they relate to; Kubernetes pods, container names, DNS names, etc. The process of augmenting
the eBPF data with these high-level concepts is called enrichment.

Enrichment flows the other way, too. Inspektor Gadget enables users to do high-performance
in-kernel filtering by only referencing high-level concepts such as Kubernetes pods, container
names, etc.; automatically translating these to the corresponding low-level kernel resources.

See the [enrichment documentation](docs/enrichment)

### What is an operator

In Inspektor Gadget, an operator is any part of the framework where an action is taken. Some operators are under-the-hood (i.e. fetching and loading Gadgets) while others are user-exposed (enrichment, filtering, export, etc.) and can be reordered and overridden.

> See the [operator documentation](docs/operators.md) for more information.

Reviewer: What do you think about this description instead https://github.com/inspektor-gadget/inspektor-gadget/blob/ece7cf9ab00bef96fed8e93caabe3e3c717da416/docs/core-concepts/_index.md#operators? 

/cc @flyth

* [Overview]()
* [Concepts]()
* [Components]()
* [Gadgets]()
* [Reference]()
* [Contributing]()

We've created a blog series that goes into several topics and concepts in details.

* [A new approach to eBPF observability]()
* [Concepts: A basic primer on Gadgets]()
* [Concepts: Advance Gadget usage]()
* [Concepts: Operators in Inspektor Gadget]()
* [Collecting metrics with Inspektor Gadget]()
* [Interactive debugging with Inspektor Gadget]()
* [Using Inspektor Gadget to leverage eBPF in your project]
* [Exploring Inspektor Gadget's security mechanisms]()

## Installation

Inspektor Gadget can be installed in several different ways depending on your environment and needs.

### On Linux

This installs the `ig` and `gadgetctl` tools

TODO: sudo dnf install `inspektor-gadget`

### For Kubernetes

Use the [krew](https://sigs.k8s.io/krew) plugin manager to install the gadget `kubectl` plugin and
deploy Inspector Gadget into a Kubernetes cluster.

```bash
kubectl krew install gadget
kubectl gadget deploy
```

### On MacOS

This installs the `gadgetctl` client tool which enables communicating with `ig` running in daemon
mode on a Linux host.

```bash

> brew install ig

Author: Ah, that should actually be `brew install inspektor-gadget` and yep, we don't have it yet. We should add a package there to make it easy for MacOS users.

## Features

* Build and package eBPF programs into OCI images called Gadgets
* Targets Kubernetes clusters and Linux hosts
* Collect and export data to observability tools with a simple command (and soon via declarative configuration)
* Security mechanisms to restrict and lock-down which Gadgets can be run
* Automatic [enrichment](#what-is-enrichment): map kernel data to high-level resources like Kubernetes and container runtimes
* Supports [WebAssembly](https://webassembly.org/) modules to post-process data and customize IG [operators](#what-is-an-operator); using any WASM-supported language 
* Supports many modes of operation; cli, client-server, API, embeddable via Golang library

## Quick start

This assumes you have the required Inspektor Gadget tools installed as outlined in the [Installation instructions](#installation).

The following examples use the [trace_open Gadget](/gadgets/trace_open/README.md) which triggers when a file is open on the system.

### Kubernetes 

Deploy and run on a Kubernetes cluster

```bash
kubectl gadget deploy
kubectl gadget run trace_open
```

Or use via `kubectl debug` (requires no installation) for a single node

```bash

> kubectl debug --profile=sysadmin node/minikube-docker -ti --image=ghcr.io/inspektor-gadget/ig -- ig trace open

Reviewer: ```suggestion
kubectl debug --profile=sysadmin node/minikube-docker -ti --image=ghcr.io/inspektor-gadget/ig -- ig run trace_open
```

Doesn't work right now but works after #3261

```

### Linux

Runs Inspektor Gadget locally on Linux.

```bash
ig run trace_open
```

### Via Docker

```bash

> docker run -ti --rm --privileged -v /:/host --pid=host ghcr.io/inspektor-gadget/ig trace open

Reviewer: ```suggestion
docker run -ti --rm --privileged -v /:/host --pid=host ghcr.io/inspektor-gadget/ig run trace_open
```

(same as above)

```

### MacOS or Windows (but not exclusively)

Run the following on a Linux machine to make `ig` available to clients.

```bash
ig daemon --host=tcp://0.0.0.0:1234
```

Use the `gadgetctl` client to connect and run commands remote Linux servers.

```bash
gadgetctl --remote-address=tcp://$IP:1234
gadgetctl run trace_open
```

***The above demonstrates the simplest command. To learn how to filter, export, etc. please consult the documentation for the [run](/docs/guides/run.md) command***.
 
## Core concepts

### What is a Gadget?

Gadgets are the central component in the Inspektor Gadget framework. A Gadget is an [OCI image](https://opencontainers.org/) that includes one or more eBPF programs, metadata YAML file and, optionally, WASM modules for post processing and a logo file.

> As OCI images, they use the same tooling as containers and share the same attributes; shareable, modular, deployable, etc.

Reviewer 2: What do you mean with tooling? You mean you can use `oras`, `crane`, etc. to manipulate them? I think it's confusing because then you talk about "deployment", so I don't want users to think that they can run Gadgets with docker 😅. 

```suggestion
As OCI images, they can be stored in a container registry (compliant with the OCI specifications), making them easy to distribute and share.
```

the eBPF data with these high-level concepts is called *enrichment*.

Enrichment flows the other way, too. Inspektor Gadget enables users to do high-performance
in-kernel filtering by only referencing high-level concepts such as Kubernetes pods, container
names, etc.; automatically translating these to the corresponding low-level kernel resources.

See the [enrichment documentation](docs/enrichment)

### What is an operator

In Inspektor Gadget, an operator is any part of the framework where an action is taken. Some operators are under-the-hood (i.e. fetching and loading Gadgets) while others are user-exposed (enrichment, filtering, export, etc.) and can be reordered and overridden.

See the [operator documentation](docs/operators.md) for more information.

### Further learning

Use the [project documentation](/docs/_index.md) to learn more about:

* [Overview](/docs/_index.md)
* [Concepts](/docs/core_concepts/_index.md)
* [Components](https://inspektor-gadget.io/docs/components)
* [Gadgets](https://inspektor-gadget.io/docs/gadgets)
* [Reference](https://inspektor-gadget.io/docs/reference)

> * [Contributing]()

Reviewer: I suppose we need to update all of this to point to the website.

Author: Yes, please.

## Installation

Inspektor Gadget can be installed in several different ways depending on your environment and needs.

### On Linux

This installs the `ig` and `gadgetctl` tools

> TODO: sudo dnf install `inspektor-gadget`

Reviewer: Remove these TODOS?

Author: Yes, but updating to the real install instructions. I didn't see where our easy instructions were.

### For Kubernetes

Use the [krew](https://sigs.k8s.io/krew) plugin manager to install the gadget `kubectl` plugin and

> deploy Inspector Gadget into a Kubernetes cluster.

Reviewer: Inspector -> Inspe*kkkkkkkkkkkkkk*tor :D

```bash
kubectl krew install gadget
kubectl gadget deploy
```

### On MacOS

This installs the `gadgetctl` client tool which enables communicating with `ig` running in daemon
mode on a Linux host.

```bash
brew install ig
```

### On Windows

This installs the `gadgetctl` client tool which enables communicating with `ig` running in daemon
mode on a Linux host.

```bash
# if using Chocolatey
choco install inspektor-gadget
# or if using WinGet

> winget install inspektor-gadget

Author: Yep, trying to point out that this is what the experience should be and if any one wants to tackle that it would be welcome. The Headlamp has experience with this.
